### DISCUSSION TITLE
digest pinning for reusable workflows currently referencing branch name within github actions

### DISCUSSION THREAD

**Discussion by travi** [renovatebot/renovate#22320] [Request Help] [answered]
### How are you running Renovate?

Mend Renovate hosted app on github.com

### If you're self-hosting Renovate, tell us what version of Renovate you run.

n/a

### If you're self-hosting Renovate, select which platform you are using.

None

### If you're self-hosting Renovate, tell us what version of the platform you run.

n/a

### Was this something which used to work for you, and then stopped?

I never saw this working

### Wanted end result.

i would like to receive a PR to switch a reusable from targeting the default branch to instead reference the sha for the latest version on the default branch. following this switch, i would like to continue to receive updates of that version, similar to pinned github actions defined within a workflow.

i do not tag my reusable workflows with semantic versions because multiple can exist in a single repository. it is somewhat awkward to reference the sha manually, so i have many repos where i started with the lazy approach of referencing by branch name instead. i would like to improve that situation, so the automation from renovate could be very helpful.

is there an existing option that i'm overlooking that could recognize the branch name and understand that i would prefer that be converted to a digest? if not has this already been considered from a feature request perspective? are there scenarios where this desire might conflict with the intended behavior of managing github actions beyond reusable workflows specifically?

### What you tried so far.

i'm already [using the helper to pin gh action digests](https://github.com/form8ion/renovate-config/blob/8894624d6b7116eb08f6d2c526c9fd1fc87976be/default.json#L7).

based on the following, i expect that continued updates would already work and i expect that semantic versions would also already be converted to digests based on my use of the helper mentioned above. since i do not add tags for semantic versions for the reusable workflows, i have not tried that option.
* https://github.com/renovatebot/renovate/issues/12212
* https://github.com/renovatebot/renovate/issues/20872

the dependency dashboard does not show matches for the usages of these reusable workflows like is shown in this reproduction repository: https://github.com/iterative/renovate-reusable-workflow-reproduction/issues/3. i expect that this is because of referencing the reusable workflow by branch name instead of semantic version or digest (`"skipReason": "invalid-version"`)

here is an example reusable workflow reference in one of my workflow files: https://github.com/form8ion/project/blob/6f707e44c6aea8a925d22d8dafed369f2392ae0f/.github/workflows/node-ci.yml#L58

### Relevant debug logs

<details><summary>Logs</summary>

```
{
            "autoReplaceStringTemplate": "{{depName}}/.github/workflows/release-package.yml@{{#if newDigest}}{{newDigest}}{{#if newValue}} # {{newValue}}{{/if}}{{/if}}{{#unless newDigest}}{{newValue}}{{/unless}}",
            "commitMessageTopic": "{{{depName}}} action",
            "currentValue": "master",
            "datasource": "github-tags",
            "depName": "form8ion/.github",
            "depType": "action",
            "packageName": "form8ion/.github",
            "replaceString": "form8ion/.github/.github/workflows/release-package.yml@master",
            "skipReason": "invalid-version",
            "versioning": "docker",
            "updates": []
          }
        ],
```

</details>

**Comment by rarkins** [comment]
Is it a github host? Is Renovate treating it as github-tags?

**Reply by travi** [reply]
I don't think I understand the question enough to follow what details you are trying to uncover, so I'll try to just be thorough. These projects are on public GitHub, use the hosted version of renovate, and reference reusable workflows hosted in a separate repo on public GitHub, sometimes across organizations. In the case of the above example, this is the referenced reusable workflow: https://github.com/form8ion/.github/blob/master/.github/workflows/release-package.yml

The log shows that renovate is using github-tags as the data source. I assume that is the default for actions, but am unsure about those details or how the pin-digests helper should impact that at this stage.

**Reply by rarkins** [reply, marked as answer]
If the datasource is `github-tags` then unfortunately there's probably no way to get this to work without quite some rework. In short.. a branch is not a tag, so that datasource won't find your "current value" (branch), let alone its digest. We'd need it to use `git-refs` instead (refs are a superset of tags and include branch names too). If you are referencing public repos then at least we wouldn't need to think about authentication for git-refs.

This type of case would be helped greatly by #19459

**Reply by travi** [reply]
Ok. Thanks sense. Appreciate the explanation.

Sounds like it is worth me exploring other options for automating the change. I'll probably just make an octoherd script for the initial switch and let renovate take over for starting up to date.

**Reply by rarkins** [reply]
Can you copy in here now the reference looks in a GitHub action ? Ie the text you're trying to pin digest for

**Reply by travi** [reply]
this is an example reference of the reusable workflow:

```yaml
uses: form8ion/.github/.github/workflows/release-package.yml@master
```

is that what you mean?

**Reply by rarkins** [reply]
Thanks. So reusable workflows always default to the current platform, or GitHub.com ?

**Reply by travi** [reply]
to be honest, i'm not sure. i've only leveraged reusable workflows where the reusable workflow and the consuming workflow are both on github.com. the docs for free, pro, & team github outline the options available for calling a reusable workflow here: https://docs.github.com/en/actions/using-workflows/reusing-workflows#calling-a-reusable-workflow

GHES would be the situation where i would expect potential need for more specificity between referencing reusable workflows on the GHES instance vs github.com. the GHES version of the same docs don't seem to clarify that details: https://docs.github.com/en/enterprise-server@3.8/actions/using-workflows/reusing-workflows#calling-a-reusable-workflow and i don't have the experience with this situation on GHES to know a better answer for this case.

is this the situation you had in mind, or are there others that you were considering?

**Reply by rarkins** [reply]
Yes, that's what I had in mind. I'm guessing reusable workflows must be on the same host, but it doesn't seem certain.

Regarding your use case, I'm not sure it will work even if you manually pin. Ultimately if it has a `@master` suffix but is sent to `github-tags` then `master` won't be found.

I'm wondering if we should instead send certain patterns to `git-refs` instead.

**Reply by travi** [reply]
> I'm not sure it will work even if you manually pin

is this because pinning to a specific SHA wouldnt match with the github-tags datasource since the SHA isnt also a tagged commit?

i got the change to sit down and try to manually pin to an older commit in one usage to test the behavior here: https://github.com/form8ion/project/commit/ff16783ecf39cac68bb0c12524f036416bcfed86

the log for that run shows that the workflow file was scanned, but shows no output at all in this case for the reusable workflow:
```
DEBUG: Matched 4 file(s) for manager github-actions: .github/workflows/codeql-analysis.yml, .github/workflows/node-ci.yml, .github/workflows/ossf-scorecard.yml, .github/workflows/release.yml
```

based on what you mentioned above, i was curious if it would match and update to the latest commit SHA or show `"skipReason": "invalid-version"` again, similar to the situation where the branch name was used as the ref. i am surprised that it was neither and there is no output for that usage. i do also notice that it is not listed as a detected dependency [in the dependency dashboard for that project](https://github.com/form8ion/project/issues/828), but i imagine that is related to the lack of showing up as a match in the log.

**Reply by travi** [reply]
for comparison, the dependency dashboard for a different project (where the same reusable workflow is used, but still using branch name as the ref) [shows the `release.yml` workflow as a matched file](https://github.com/form8ion/javascript/issues/13), but containing no detected dependencies. in the dashboard mentioned above, the `release.yml` workflow is not even listed as a matched file, even though the log shows that it was matched. 

i'm not sure which behavior is expected for these cases, but they are at least different in this situation.

**Reply by rarkins** [reply]
I was assuming that you had already confirmed reusable workflows work as expected when using a tag. If they don't work at all, then that sounds like a pre-condition

**Reply by travi** [reply]
I haven't personally confirmed because I don't have any reusable workflows that I am assigning tags to. I linked to a couple issues in my original post above that I thought confirmed that they are working with tags. If we want more confirmation than that, I can set up a test repo when I'm at a computer next.

**Reply by rarkins** [reply]
Maybe it just fails to recognize your dependency at all due to the branch syntax
